### ISSUE TITLE
Files missing in nodebb/public

### ISSUE THREAD

**Issue by vstoykovbg** [NodeBB/NodeBB#5452]
My main point here is: it is not efficient static files to be served by NodeBB (instead of Nginx directly).

I have configured Nginx to first look for files on `nodebb/public` directory before sending requests to NodeBB:

```
    location ~ (^/(assets|images|language|less|plugins|sounds|templates|uploads|vendor)|\.(js|css|ico|png)$) {
        root /home/example/nodebb/public;
        try_files $uri @nodebb;
    }
```
The problem is that not all static files can be found there.

I made a dirty workaround with this `.bash` script:
```
#!/bin/bash

if [ ! -f ./nodebb ] || [ ! -d ./public ]; then
    echo "We are not in nodebb directory. Exiting."
    exit;
fi

mkdir -p ./public/assets/
mkdir -p ./public/src/modules/composer/
mkdir -p ./public/plugins/nodebb-plugin-markdown/styles/

cp ./public/vendor/buzz/buzz.js ./public/src/modules/buzz.js
cp ./node_modules/mousetrap/mousetrap.js ./public/src/modules/mousetrap.js

cp ./node_modules/nodebb-plugin-composer-default/static/lib/composer.js ./public/src/modules/composer.js

cp ./node_modules/nodebb-plugin-composer-default/static/lib/composer/*.js ./public/src/modules/composer/

cp ./node_modules/nodebb-plugin-markdown/public/styles/* ./public/plugins/nodebb-plugin-markdown/styles/

cp ./node_modules/nodebb-plugin-markdown/public/js/highlight.js ./public/src/modules/highlight.js

cp -r ./build/public/language ./public/assets/
```
It should be executed from the `nodebb` directory:

```
$ cd /home/user/nodebb/
$ ./dirty.fix.bash
```

However, this is not an elegant solution and this script should be executed every time when NodeBB is  updated.

Also the regular expression in `nginx.conf` is not as efficient as possible. It would be more efficient if all static files are moved within `nodebb/public/static` directory.

This way `nginx.conf` can be more efficient:

```
    location ^~ /static {
        root /home/example/nodebb/public;
        try_files $uri @nodebb;
    }
```

Also, I suggest robots.txt  to be static file, not served by NodeBB (created by NodeBB build script):

```
    location = /robots.txt {
        root /home/example/nodebb/public;
        try_files $uri =404;
    }

  #  Some browser may ask for /favicon.ico. I tested it only with Firefox and Chromium
  #  and they take into consideration the `rel="icon"` link.
  # Maybe it is safe to delete the following location block if the `rel="icon"` link points to the
  # `/static` directory
    location = /favicon.ico {
        root /home/example/nodebb/public;
        try_files $uri =404;
    }
```

**Comment by julianlam** [comment]
Hi @vstoykovbg -- can you believe that thanks to @pitaj, this is now a solved problem?

Check out the write-up here: https://blog.nodebb.org/introducing-the-build-system-in-v1-4-3/

tl;dr -- everything is saved in `/build/public` now, and you can set up nginx thusly:

```
location ~ ^/assets/(.*) {  
  root /path/to/nodebb/;
  try_files /build/public/$1 /public/$1 @nodebb;
}

location /plugins/ {  
  root /path/to/nodebb/build/public/;
  try_files $uri @nodebb;
}
```

**Comment by pitaj** [comment]
We moved built files to `build/public`. Check out this blog post:

https://blog.nodebb.org/introducing-the-build-system-in-v1-4-3/

**Comment by vstoykovbg** [comment]
I tested it with the latest version (`master`) and it works.

First, it did not worked:
```
2017/02/16 17:58:17 [crit] 12093#0: *939 stat() "/home/example/nodebb/build/public/src/modules/composer/drafts.js" failed (13: Permission denied), client: 192.168.100.42, server: example.com, request: "GET /assets/src/modules/composer/drafts.js HTTP/1.1", host: "example.com"
```

But I fixed it. The problem was SELinux - symlinks was disabled.

```
# grep nginx /var/log/audit/audit.log | audit2allow -m local

module local 1.0;

require {
	type httpd_t;
	type user_home_t;
	class lnk_file read;
}

#============= httpd_t ==============

#!!!! This avc can be allowed using the boolean 'httpd_enable_homedirs'
allow httpd_t user_home_t:lnk_file read;
```

I added this to my previous SELinux module:

```
# echo "module nginxfix 2.0;

require {
	type httpd_t;
	type var_t;
	type initrc_t;
	class sock_file write;
	class unix_stream_socket connectto;
	type user_home_t;
	class lnk_file read;
}



#============= httpd_t ==============

allow httpd_t initrc_t:unix_stream_socket connectto;
allow httpd_t var_t:sock_file write;
allow httpd_t user_home_t:lnk_file read;"  > /tmp/nginxfix.te


# checkmodule -M -m -o  /tmp/nginxfix.mod  /tmp/nginxfix.te
# semodule_package -o  /tmp/nginxfix.pp -m  /tmp/nginxfix.mod
# semodule -i  /tmp/nginxfix.pp
```

This module is enabling socket files and symlinks.

Also I have these SELinux settings:

Allowing logs in `/var/log/virtualmin(/.*)?`:
```
# semanage fcontext -a -t httpd_log_t "/var/log/virtualmin(/.*)?" 
```
Allowing Nginx to access content on `/home/*` directories:

```
# setsebool -P httpd_read_user_content on
```

Allowing `proxy_pass http://io_nodes;`:

```
# setsebool -P httpd_can_network_connect on
```

**Comment by pitaj** [comment]
> symlinks were disabled

That would do it
